src/libpsc/tests/test_bnd_prt_inflow.cxx

```diff
@@ -0,0 +1,425 @@
+  EXPECT_EQ(prt.u, u);
+  EXPECT_EQ(prt.w, wni);
+  EXPECT_EQ(prt.kind, 0);
+  EXPECT_EQ(prt.tag, tag);
+}
+
+template <typename MP, typename DIM>
+class Inflow
+{
+public:
+  using Mparticles = MP;
+  using real_t = typename Mparticles::real_t;
+  using Dim = DIM;
+
+  // offset_in_cell: () -> double
+  Inflow(const Grid_t& grid, psc_particle_npt npt,
+         double (*offset_in_cell_dist)())
+    : grid_(grid),
+      setup_particles_(grid),
+      advance_(grid.dt),
+      // FIXME np.p is a std::function and is called many times; better to use a
+      // lambda
+      np_{npt.kind, npt.n, setup_particles_.createMaxwellian(npt)},
+      offset_in_cell_dist_(offset_in_cell_dist),
+      inj_dim_idx_(!Dim::InvarX::value   ? 0
+                   : !Dim::InvarY::value ? 1
+                   : !Dim::InvarZ::value ? 2
+                                         : -1)
+  {
+    assert(inj_dim_idx_ >= 0);
+  }
+
+  auto get_advanced_prt(Double3 pos, real_t wni)
+  {
+    auto prt = setup_particles_.setupParticle(np_, pos, wni);
+
+    auto v = advance_.calc_v(prt.u);
+    advance_.push_x(prt.x, v, 1.0);
+
+    return prt;
+  }
+
+  template <typename Injector>
+  void inject_into_boundary_cell(Injector& injector,
+                                 Int3 boundary_cell_global_idx)
+  {
+    assert(boundary_cell_global_idx[inj_dim_idx_] == 0);
+    boundary_cell_global_idx[inj_dim_idx_] = -1;
+
+    int n_in_cell = setup_particles_.get_n_in_cell(np_);
+    double wni = setup_particles_.getWeight(np_, n_in_cell);
+
+    for (int cnt = 0; cnt < n_in_cell; cnt++) {
+      Double3 offset = {offset_in_cell_dist_(), offset_in_cell_dist_(),
+                        offset_in_cell_dist_()};
+      auto pos =
+        (Double3(boundary_cell_global_idx) + offset) * grid_.domain.dx +
+        grid_.domain.corner;
+      auto prt = get_advanced_prt(pos, wni);
+
+      if (prt.x[inj_dim_idx_] < grid_.domain.corner[inj_dim_idx_]) {
+        continue;
+      }
+
+      injector(prt);
+    }
+  }
+
+  template <typename Injector>
+  void inject_into_boundary_patch(Injector& injector,
+                                  const Grid_t::Patch& boundary_patch)
+  {
+    Int3 ilo = boundary_patch.off;
+    Int3 ihi = ilo + grid_.ldims;
+
+    assert(ilo[inj_dim_idx_] == 0);
+
+    int dim1 = std::min((inj_dim_idx_ + 1) % 3, (inj_dim_idx_ + 2) % 3);
+    int dim2 = std::max((inj_dim_idx_ + 1) % 3, (inj_dim_idx_ + 2) % 3);
+
+    for (Int3 cell_idx = ilo; cell_idx[dim1] < ihi[dim1]; cell_idx[dim1]++) {
+      for (cell_idx[dim2] = ilo[dim2]; cell_idx[dim2] < ihi[dim2];
+           cell_idx[dim2]++) {
+        inject_into_boundary_cell(injector, cell_idx);
+      }
+    }
+  }
+
+  template <typename Injectors>
+  void inject_into_boundary(Injectors& injectors_by_patch)
+  {
+    for (int patch_idx = 0; patch_idx < grid_.n_patches(); patch_idx++) {
+      const auto& patch = grid_.patches[patch_idx];
+
+      if (patch.off[inj_dim_idx_] != 0) {
+        continue;
+      }
```

Comment:
There is a `at_boundary_lo()` or something of that sort somewhere, but not sure it's easily reusable here.

Reply:
I see `mrc_domain_at_boundary_lo`, but yeah, doesn't seem usable here